In [ ]:
import pandas as pd
from pathlib import Path

PROCESSED_DIR = Path(r"D:\1 Univesrity work\Lect\Data  management\US Retail Sales and Public Holidays Analysis\DATA\processed")

orders_path = PROCESSED_DIR / "orders_clean.csv"
holidays_path = PROCESSED_DIR / "holidays_clean.csv"

orders = pd.read_csv(orders_path, parse_dates=['Order Date'])
holidays = pd.read_csv(holidays_path, parse_dates=['date'])

# Quick checking the data
print("Orders shape:", orders.shape)
print("Holidays shape:", holidays.shape)
print(orders.head())
print(holidays.head())

Orders shape: (9994, 25)
Holidays shape: (52, 3)
   Row ID        Order ID Order Date   Ship Date       Ship Mode Customer ID  \
0       6  CA-2014-115812 2014-06-09  2014-06-14  Standard Class    BH-11710   
1       7  CA-2014-115812 2014-06-09  2014-06-14  Standard Class    BH-11710   
2       8  CA-2014-115812 2014-06-09  2014-06-14  Standard Class    BH-11710   
3       9  CA-2014-115812 2014-06-09  2014-06-14  Standard Class    BH-11710   
4      10  CA-2014-115812 2014-06-09  2014-06-14  Standard Class    BH-11710   

     Customer Name   Segment        Country         City  ... Sub-Category  \
0  Brosina Hoffman  Consumer  United States  Los Angeles  ...  Furnishings   
1  Brosina Hoffman  Consumer  United States  Los Angeles  ...          Art   
2  Brosina Hoffman  Consumer  United States  Los Angeles  ...       Phones   
3  Brosina Hoffman  Consumer  United States  Los Angeles  ...      Binders   
4  Brosina Hoffman  Consumer  United States  Los Angeles  ...   Appliances   

 

In [ ]:
import pandas as pd

# Ensure 'Order Date' and 'date' columns are datetime (just in case)
orders['Order Date'] = pd.to_datetime(orders['Order Date'])
holidays['date'] = pd.to_datetime(holidays['date'])

# Merge orders with holidays on date
merged_df = orders.merge(
    holidays[['date', 'name']], 
    how='left',
    left_on='Order Date',
    right_on='date'
)

# Drop the extra date
merged_df = merged_df.drop(columns=['date'])

# Create a boolean column: True if it's a holiday, False otherwise
merged_df['IsHoliday'] = ~merged_df['name'].isna()

merged_df = merged_df.rename(columns={'name': 'Holiday Name'})

# Quick check
print(merged_df[['Order ID', 'Order Date', 'Holiday Name', 'IsHoliday']].head(10))

         Order ID Order Date  Holiday Name  IsHoliday
0  CA-2014-115812 2014-06-09           NaN      False
1  CA-2014-115812 2014-06-09           NaN      False
2  CA-2014-115812 2014-06-09           NaN      False
3  CA-2014-115812 2014-06-09           NaN      False
4  CA-2014-115812 2014-06-09           NaN      False
5  CA-2014-115812 2014-06-09           NaN      False
6  CA-2014-115812 2014-06-09           NaN      False
7  CA-2014-105893 2014-11-11  Veterans Day       True
8  CA-2014-167164 2014-05-13           NaN      False
9  CA-2014-143336 2014-08-27           NaN      False


In [6]:
merged_df

Row ID        Order ID Order Date   Ship Date       Ship Mode  \
0          6  CA-2014-115812 2014-06-09  2014-06-14  Standard Class   
1          7  CA-2014-115812 2014-06-09  2014-06-14  Standard Class   
2          8  CA-2014-115812 2014-06-09  2014-06-14  Standard Class   
3          9  CA-2014-115812 2014-06-09  2014-06-14  Standard Class   
4         10  CA-2014-115812 2014-06-09  2014-06-14  Standard Class   
...      ...             ...        ...         ...             ...   
9989    9989  CA-2017-163629 2017-11-17  2017-11-21  Standard Class   
9990    9991  CA-2017-121258 2017-02-26  2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26  2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26  2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04  2017-05-09    Second Class   

     Customer ID    Customer Name    Segment        Country         City  ...  \
0       BH-11710  Brosina Hoffman   Consumer  United States  Los Angeles  ...   
1       BH-11710  Brosina Hoffman   Consumer  United States  Los Angeles  ...   
2       BH-11710  Brosina Hoffman   Consumer  United States  Los Angeles  ...   
3       BH-11710  Brosina Hoffman   Consumer  United States  Los Angeles  ...   
4       BH-11710  Brosina Hoffman   Consumer  United States  Los Angeles  ...   
...          ...              ...        ...            ...          ...  ...   
9989    RA-19885     Ruben Ausman  Corporate  United States       Athens  ...   
9990    DB-13060      Dave Brooks   Consumer  United States   Costa Mesa  ...   
9991    DB-13060      Dave Brooks   Consumer  United States   Costa Mesa  ...   
9992    DB-13060      Dave Brooks   Consumer  United States   Costa Mesa  ...   
9993    CC-12220     Chris Cortes   Consumer  United States  Westminster  ...   

        Sales  Quantity Discount Discount value   Profit      COGS  \
0      48.860         7      0.0         0.0000  14.1694  -34.6906   
1       7.280         4      0.0         0.0000   1.9656   -5.3144   
2     907.152         6      0.2      -181.4304  90.7152 -635.0064   
3      18.504         3      0.2        -3.7008   5.7825   -9.0207   
4     114.900         5      0.0         0.0000  34.4700  -80.4300   
...       ...       ...      ...            ...      ...       ...   
9989  206.100         5      0.0         0.0000  55.6470 -150.4530   
9990   91.960         2      0.0         0.0000  15.6332  -76.3268   
9991  258.576         2      0.2       -51.7152  19.3932 -187.4676   
9992   29.600         4      0.0         0.0000  13.3200  -16.2800   
9993  243.160         2      0.0         0.0000  72.9480 -170.2120   

     Shipping time range  Year sales  Holiday Name  IsHoliday  
0                      5        2014           NaN      False  
1                      5        2014           NaN      False  
2                      5        2014           NaN      False  
3                      5        2014           NaN      False  
4                      5        2014           NaN      False  
...                  ...         ...           ...        ...  
9989                   4        2017           NaN      False  
9990                   5        2017           NaN      False  
9991                   5        2017           NaN      False  
9992                   5        2017           NaN      False  
9993                   5        2017           NaN      False  

[9994 rows x 27 columns]

In [8]:
print(merged_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Row ID               9994 non-null   int64         
 1   Order ID             9994 non-null   object        
 2   Order Date           9994 non-null   datetime64[ns]
 3   Ship Date            9994 non-null   object        
 4   Ship Mode            9994 non-null   object        
 5   Customer ID          9994 non-null   object        
 6   Customer Name        9994 non-null   object        
 7   Segment              9994 non-null   object        
 8   Country              9994 non-null   object        
 9   City                 9994 non-null   object        
 10  State                9994 non-null   object        
 11  Postal Code          9994 non-null   int64         
 12  Region               9994 non-null   object        
 13  Product ID           9994 non-nul

In [9]:
# Check duplicates
duplicates = merged_df.duplicated().sum()
print(f"Total duplicate rows after merge: {duplicates}")

Total duplicate rows after merge: 0


In [10]:
# Check nulls
nulls = merged_df.isnull().sum()
print("Null values per column after merge:\n", nulls)

Null values per column after merge:
 Row ID                    0
Order ID                  0
Order Date                0
Ship Date                 0
Ship Mode                 0
Customer ID               0
Customer Name             0
Segment                   0
Country                   0
City                      0
State                     0
Postal Code               0
Region                    0
Product ID                0
Category                  0
Sub-Category              0
Product Name              0
Sales                     0
Quantity                  0
Discount                  0
Discount value            0
Profit                    0
COGS                      0
Shipping time range       0
Year sales                0
Holiday Name           9544
IsHoliday                 0
dtype: int64


In [ ]:
merged_df['Holiday Name'] = merged_df['Holiday Name'].fillna("No Holiday")

In [12]:
# Check nulls
nulls = merged_df.isnull().sum()
print("Null values per column after merge:\n", nulls)

Null values per column after merge:
 Row ID                 0
Order ID               0
Order Date             0
Ship Date              0
Ship Mode              0
Customer ID            0
Customer Name          0
Segment                0
Country                0
City                   0
State                  0
Postal Code            0
Region                 0
Product ID             0
Category               0
Sub-Category           0
Product Name           0
Sales                  0
Quantity               0
Discount               0
Discount value         0
Profit                 0
COGS                   0
Shipping time range    0
Year sales             0
Holiday Name           0
IsHoliday              0
dtype: int64


Commit: Integrate Orders with Holiday API

- Loaded cleaned orders and holidays datasets.
- Converted 'Order Date' and 'date' columns to datetime for consistency.
- Merged orders with holidays on order date using a left join.
- Added 'Holiday Name' column with actual holiday names where applicable.
- Created 'IsHoliday' boolean flag: True if order falls on a holiday, False otherwise.
- Verified integration: 
    - Total orders: 9994
    - Orders on holidays: 450
    - Orders not on holidays: 9544
    - Percentage of orders on holidays: 4.50%
- Confirmed no integration errors; unmatched orders are genuine non-holidays.
- Dataset is ready for feature engineering and analysis.

In [13]:
# Total orders
total_orders = len(merged_df)

# Orders that matched a holiday
orders_on_holidays = merged_df['IsHoliday'].sum()

# Orders that did NOT match a holiday
orders_not_on_holidays = total_orders - orders_on_holidays

# Percentage matched
matched_pct = orders_on_holidays / total_orders * 100

print(f"Total orders: {total_orders}")
print(f"Orders on holidays: {orders_on_holidays}")
print(f"Orders not on holidays: {orders_not_on_holidays}")
print(f"Percentage of orders correctly integrated with holidays: {matched_pct:.2f}%")

Total orders: 9994
Orders on holidays: 450
Orders not on holidays: 9544
Percentage of orders correctly integrated with holidays: 4.50%


In [20]:
merged_df.to_csv(PROCESSED_DIR / "orders_integrated_holidays.csv", index=False)
